aquadoggo/src/graphql/schema_builders/document.rs

```diff
@@ -0,0 +1,474 @@
+
+use async_graphql::dynamic::{Field, FieldFuture, InputValue, Object, TypeRef};
+use dynamic_graphql::{Error, FieldValue, ScalarValue};
+use log::debug;
+use p2panda_rs::storage_provider::traits::DocumentStore;
+use p2panda_rs::{document::traits::AsDocument, schema::Schema};
+
+use crate::db::SqlStore;
+use crate::graphql::scalars::{DocumentIdScalar, DocumentViewIdScalar};
+use crate::graphql::types::DocumentMeta;
+use crate::graphql::utils::{downcast_id_params, fields_name, get_document_from_params};
+
+const QUERY_ALL_PREFIX: &str = "all_";
+const FIELDS_FIELD: &str = "fields";
+const META_FIELD: &str = "meta";
+const DOCUMENT_META_SCHEMA_ID: &str = "DocumentMeta";
+const DOCUMENT_ID_ARG: &str = "id";
+const DOCUMENT_VIEW_ID_ARG: &str = "viewId";
+const DOCUMENT_ID_SCALAR: &str = "DocumentId";
+const DOCUMENT_VIEW_ID_SCALAR: &str = "DocumentViewId";
+
+/// Build a graphql object type for a p2panda schema.
+///
+/// Contains resolvers for both `fields` and `meta`. The former simply passes up the query
+/// arguments to it's children query fields. The latter retrieves the document being queried and
+/// already constructs and returns the `DocumentMeta` object.
+pub fn build_document_schema(schema: &Schema) -> Object {
+    let document_fields_name = fields_name(schema.id());
+    Object::new(schema.id().to_string())
+        // The `fields` of a document, passes up the query arguments to it's children.
+        .field(Field::new(
+            FIELDS_FIELD,
+            TypeRef::named(document_fields_name),
+            move |ctx| {
+                FieldFuture::new(async move {
+                    // Here we just pass up the root query parameters to be used in the fields resolver.
+                    let params = downcast_id_params(&ctx);
+                    Ok(Some(FieldValue::owned_any(params)))
+                })
+            },
+        ))
+        // The `meta` field of a document, resolves the `DocumentMeta` object.
+        .field(Field::new(
+            META_FIELD,
+            TypeRef::named(DOCUMENT_META_SCHEMA_ID),
+            move |ctx| {
+                FieldFuture::new(async move {
+                    let store = ctx.data_unchecked::<SqlStore>();
+
+                    // Downcast the parameters passed up from the parent query field.
+                    let (document_id, document_view_id) = downcast_id_params(&ctx);
+                    // Get the whole document.
+                    let document =
+                        get_document_from_params(store, &document_id, &document_view_id).await?;
+
+                    // Construct `DocumentMeta` and return it. We defined the document meta
+                    // type and already registered it in the schema. It's derived resolvers
+                    // will handle field selection.
+                    //
+                    // TODO: We could again optimize here by defining our own resolver logic
+                    // for each field.
```

Comment:
I'm not sure if I understand this optimization idea. Do you suggest to have an individual SQL query for _each_ field? That wouldn't be optimal probably, or do you suggest to pass down the required list of fields to some sort of resolver which then builds _one_ custom SQL query based on the requested fields?